src/coreclr/nativeaot/Runtime/windows/CoffNativeCodeManager.cpp

```diff
@@ -341,6 +340,39 @@ PTR_VOID CoffNativeCodeManager::GetFramePointer(MethodInfo *   pMethInfo,
     return NULL;
 }
 
+#ifdef TARGET_X86
+uintptr_t CoffNativeCodeManager::GetResumeSp(MethodInfo *   pMethodInfo,
+                                             REGDISPLAY *   pRegisterSet)
+{
+    PTR_uint8_t gcInfo;
+    uint32_t codeOffset = GetCodeOffset(pMethodInfo, (PTR_VOID)pRegisterSet->IP, &gcInfo);
+
+    hdrInfo infoBuf;
+    size_t infoSize = DecodeGCHdrInfo(GCInfoToken(gcInfo), codeOffset, &infoBuf);
+    PTR_CBYTE table = gcInfo + infoSize;
+
+    _ASSERTE(infoBuf.epilogOffs == hdrInfo::NOT_IN_EPILOG && infoBuf.prologOffs == hdrInfo::NOT_IN_PROLOG);
+
+    bool isESPFrame = !infoBuf.ebpFrame && !infoBuf.doubleAlign;
+
+    CoffNativeMethodInfo * pNativeMethodInfo = (CoffNativeMethodInfo *)pMethodInfo;
+    if (pNativeMethodInfo->mainRuntimeFunction != pNativeMethodInfo->runtimeFunction)
+    {
+        // Treat funclet's frame as ESP frame
+        isESPFrame = true;
+    }
+
+    if (isESPFrame)
+    {
+        const uintptr_t curESP = pRegisterSet->SP;
+        return curESP + GetPushedArgSize(&infoBuf, table, codeOffset);
+    }
+
+    const uintptr_t curEBP = pRegisterSet->GetFP();
+    return curEBP - infoBuf.stackSize + sizeof(int);
```

Comment:
This bit is tricky and may be downright incorrect in some situations. Original code uses `GetOutermostBaseFP`. In code without `localloc` it does the identical thing. In code with `localloc` it tries to read a saved ESP value from a slot that doesn't exist in NativeAOT ABI.

Reply:
I guess this means that the slot needs to be added for NativeAOT on x86?

Reply:
Maybe. I am not entirely sure it’s necessary but I am not familiar enough with the CLR ABI yet. I’ll experiment with it a bit but wanted to mention it to get the code more scrutiny.

Reply:
For posterity:
> The establisher frame definition differs between CoreRT ABI and CoreCLR ABI. In CoreRT ABI created by `@russellhadley` and `@smosier` while back, the establisher frame is always the SP value at the point that control left the frame in which execution would resume after a catch handler completes. This mismatch leads to crashes when PSPSym is recomputed in funclets of methods with localloc.

Reply:
https://github.com/dotnet/coreclr/pull/8319

Reply:
This calculation is indeed incorrect in presence of `localloc`. There are couple of options:
1) Reinstate `lvaLocAllocSPvar` on x86 for NativeAOT ABI.
2) Ignore the EBP case and leave the outgoing arguments on the stack. This can lead to unbounded stack growth in the exceptional cases. In this case the codegen uses EBP-based offset to access locals and restores the stack pointer in the epilog which mostly alleviates the other issues.
3) Encode the arg table in GC info so we can use `GetPushedArgSize` like in ESP-based frames.

There's already one other code branch that depends on `locAllocSPvar` and which would give incorrect results:
https://github.com/dotnet/runtime/blob/21f23f91a0c29d91f69013d561e82bd0960c95e1/src/coreclr/vm/gc_unwind_x86.inl#L3391-L3396

Given these considerations I am heavily leaning towards option 1.

Reply:
I implemented the option 3. Marking this as resolved.